code/modules/security/workshop.dm

```diff
@@ -0,0 +1,149 @@
+// Workshop area for the prison wing in Brig
+
+
+/obj/machinery/computer/holodeck/prison
+	name = "workshop control console"
+	desc = "a computer used to control the workshop in the prison"
+
+	mapped_start_area = /area/holodeck/prison
+	linked = /area/holodeck/prison //linked area
+	program_type = /datum/map_template/holodeck/prison //load workshop programs
+	req_access = list(ACCESS_SECURITY)
+	var/startup
+	var/offline = FALSE
+
+/obj/machinery/computer/holodeck/prison/LateInitialize()
+	var/area/computer_area = get_area(src)
+	if(istype(computer_area, /area/holodeck/prison))
+		log_mapping("Holodeck computer cannot be in a holodeck, This would cause circular power dependency.")
+		qdel(src)
+		return
+	else
+		offline_program = pick("donut", "plush")
+		. = ..()
+
+/obj/machinery/computer/holodeck/prison/generate_program_list()
+	for(var/typekey in subtypesof(program_type))
+		var/datum/map_template/holodeck/program = typekey
+		var/list/info_this = list("id" = initial(program.template_id), "name" = initial(program.name))
+		if(!(initial(program.template_id) == "offline"))
+			LAZYADD(program_cache, list(info_this))
+
+/datum/map_template/holodeck/prison/update_blacklist(turf/placement, list/input_blacklist)
+	. = ..()
+
+/obj/machinery/computer/holodeck/prison/process(delta_time=2) //don't derez items that leave the area
+	if(damaged && DT_PROB(10, delta_time))
+		for(var/turf/holo_turf in linked)
+			if(DT_PROB(5, delta_time))
+				do_sparks(2, 1, holo_turf)
+				return
+	. = ..()
+	if(!. || program == offline_program)//we dont need to scan the holodeck if the holodeck is offline
+		return
+
+	if(!floorcheck()) //if any turfs in the floor of the holodeck are broken
+		emergency_shutdown()
+		damaged = TRUE
+		visible_message("The holodeck overloads!")
+		for(var/turf/holo_turf in linked)
+			if(DT_PROB(30, delta_time))
+				do_sparks(2, 1, holo_turf)
+			SSexplosions.lowturf += holo_turf
+			holo_turf.hotspot_expose(1000,500,1)
+
+	for(var/obj/effect/holodeck_effect/holo_effect as anything in effects)
+		holo_effect.tick()
+	active_power_usage = 50 + spawned.len * 3 + effects.len * 5
```

Comment:
This behaviour is duplicated from ..() and will be run twice because it will call the parent and run the same code.